File: compute.cpp
#include <iostream>
#include <iomanip>
#include <fstream>
#include <cmath>
#include <string>
#include <sstream>

using namespace std;

// load data file
void load_initial_state(string, double *, int);
// compute dt
double update_dt(const double *, const double *,
        const double *,double, int);
// Copy function
void cpy_to(double *, double *, int);
// convert 1D list to 2D array
int to_idx(int, int, int);
// impose boundary conditions
void enforce_BC(double *, double *, double *, int);
// perform one time step computation
void time_step(double *, double *, double *,
               const double *, const double *,
               double *, double *, double *,
               double, double, int);

float const g = 127267.2; //Gravity, 9.82*(3.6)^2*1000 in [km / hr^2]

int main(){
  // Basic Parameters of the simulation :
  int      Size    = 500;  // Size of map, Size*Size [km]
  size_t   nx      = 2001; //Number of cells in each direction on the grid
  float    Tend    = 0.2;  //Simulation time in hours [hr]
  double   dx      = ((float)Size)/((float)nx); //Grid spacening
  cout << "dx = " << dx << endl;
  int numElements = nx*nx;
  size_t memsize = numElements * sizeof(double);

  // Data filename :
  ostringstream sfilename;
  string addpath = "data/";
  sfilename <<addpath<<"Data_nx"<<to_string(nx)<<"_"<<to_string(Size)<<"km_T"<<Tend<< setprecision(2);
  string filename = sfilename.str();

  // Allocate memory for Computing
  cout <<" Allocating memory .."<<endl;
  double *H, *HU, *HV, *Zdx, *Zdy, *Ht, *HUt, *HVt;
  H = (double *)malloc(memsize);
  HU = (double *)malloc(memsize);
  HV = (double *)malloc(memsize);
  Zdx = (double *)malloc(memsize);
  Zdy = (double *)malloc(memsize);
  Ht = (double *)malloc(memsize);
  HUt = (double *)malloc(memsize);
  HVt = (double *)malloc(memsize);

  // Load initial condition from data files
  cout <<" Loading data from :" << endl;
  cout <<" "+filename+"_h.bin"+" .."<<  endl;
  load_initial_state(filename+"_h.bin",H,numElements);
  cout <<" "+filename+"_hu.bin"+" .."<<  endl;
  load_initial_state(filename+"_hu.bin",HU,numElements);
  cout <<" "+filename+"_hv.bin"+" .."<<  endl;
  load_initial_state(filename+"_hv.bin",HV,numElements);
  // Load topography slopes from data files
  cout <<" "+filename+"_Zdx.bin"+" .."<<  endl;
  load_initial_state(filename+"_Zdx.bin",Zdx,numElements);
  cout <<" "+filename+"_Zdy.bin"+" .."<<  endl;
  load_initial_state(filename+"_Zdy.bin",Zdy,numElements);

  // Compute the time-step length
  double T = 0.0;
  int nt = 0;
  double dt = 0.;
  double C = 0.0;

  // Evolution loop
  while (nt < 1) {
    dt = update_dt(H,HU,HV,dx,numElements);
    cout <<"dt = "<< dt<<endl;
    if(T+dt > Tend){
      dt = Tend-T;
    }
    //Print status
    cout << "Computing for T = " << T+dt << " ("<< 100*(T+dt)/Tend << "%)"<<endl;

    // Copy solution to temp storage and enforce boundary condition
    cpy_to(Ht,H,numElements);
    cpy_to(HUt,HU,numElements);
    cpy_to(HVt,HV,numElements);
    cout<<HUt[(nx-1)*(nx-1)]<<endl;
    enforce_BC(Ht, HUt, HVt, nx);

    C = (.5*dt/dx);
    time_step(H,HU,HV,Zdx,Zdy,Ht,HUt,HVt,C,dt,nx);

    T = T + dt;
    nt++;
  }

  // Free memory space
  free(H); free(HU); free(HV); free(Zdx); free(Zdy);
  free(Ht); free(HUt); free(HVt);

  return 0;
}


void load_initial_state(string filename, double * H, int numElements){
  ifstream fin;
  fin.open(filename, ios::in|ios::binary);
  if(!fin){
    cout<<" Error, couldn't find file : "<<filename<<endl;
  }
  fin.read(reinterpret_cast<char*>(&H[0]), numElements*sizeof(double));
  fin.close();
}

double update_dt(const double *H, const double *HU,
      const double *HV, double dx, int numElements){
  //Compute the max of mu and give dt back
  double mu = 0.0;
  double newmu = 0.0;
  for(int i=0; i<numElements; i++){
    newmu = sqrt(pow(max(abs(HU[i]/H[i]-sqrt(H[i]*g)),abs(HU[i]/H[i]+sqrt(H[i]*g))),2)
                +pow(max(abs(HV[i]/H[i]-sqrt(H[i]*g)),abs(HV[i]/H[i]+sqrt(H[i]*g))),2));
    if(newmu > mu){
      mu = newmu;
      }
    }
    return dx/(sqrt(2.0)*mu);
  }

void cpy_to(double *target, double *source, int numElements){
  for(int i=0; i<numElements; i++){
    target[i] = source[i];
  }
}

int to_idx(int x, int y, int nx){
  return x * (nx-1) + y;
}

void enforce_BC(double *Ht, double *HUt, double *HVt, int nx){
  for(int i=0; i<nx; ++i){
    Ht[to_idx(0,i,nx)]  = Ht[to_idx(1,i,nx)];
    Ht[to_idx(nx-1,i,nx)] = Ht[to_idx(nx-2,i,nx)];
    Ht[to_idx(i,0,nx)]  = Ht[to_idx(i,1,nx)];
    Ht[to_idx(i,nx-1,nx)] = Ht[to_idx(i,nx-2,nx)];

    HUt[to_idx(0,i,nx)]  = HUt[to_idx(1,i,nx)];
    HUt[to_idx(nx-1,i,nx)] = HUt[to_idx(nx-2,i,nx)];
    HUt[to_idx(i,0,nx)]  = HUt[to_idx(i,1,nx)];
    HUt[to_idx(i,nx-1,nx)] = HUt[to_idx(i,nx-2,nx)];

    HVt[to_idx(0,i,nx)]  = HVt[to_idx(1,i,nx)];
    HVt[to_idx(nx,i,nx)] = HVt[to_idx(nx-1,i,nx)];
    HVt[to_idx(i,0,nx)]  = HVt[to_idx(i,1,nx)];
    HVt[to_idx(i,nx-1,nx)] = HVt[to_idx(i,nx-2,nx)];
  }
}

void time_step(double *H, double *HU, double *HV,
               const double *Zdx, const double *Zdy,
               double *Ht, double *HUt, double *HVt,
               double C, double dt, int nx){
  for(int x=1; x<nx-1; x++){
    for(int y=1; y<nx-1; y++){
      H[to_idx(x,y,nx)]   = 0.25*( Ht[to_idx(x+1,y,nx)]+Ht[to_idx(x-1,y,nx)]
                                  +Ht[to_idx(x,y+1,nx)]+Ht[to_idx(x,y-1,nx)])
                          + C   *( HUt[to_idx(x,y-1,nx)]-HUt[to_idx(x,y+1,nx)]
                                  +HVt[to_idx(x-1,y,nx)]-HVt[to_idx(x+1,y,nx)]);

      HU[to_idx(x,y,nx)]  = 0.25*( HUt[to_idx(x+1,y,nx)]+HUt[to_idx(x-1,y,nx)]
                                  +HUt[to_idx(x,y+1,nx)]+HUt[to_idx(x,y-1,nx)])
                                  -dt*g*H[to_idx(x,y,nx)]*Zdx[to_idx(x,y,nx)]
        + C*( pow(HUt[to_idx(x,y-1,nx)],2)/Ht[to_idx(x,y-1,nx)]
             +0.5*g*pow(Ht[to_idx(x,y-1,nx)],2)
             -pow(HUt[to_idx(x,y+1,nx)],2)/Ht[to_idx(x,y+1,nx)]
             -0.5*g*pow(Ht[to_idx(x,y+1,nx)],2))
        + C*( HUt[to_idx(x-1,y,nx)]*HVt[to_idx(x-1,y,nx)]/Ht[to_idx(x-1,y,nx)]
             -HUt[to_idx(x+1,y,nx)]*HVt[to_idx(x+1,y,nx)]/Ht[to_idx(x+1,y,nx)]);

      HV[to_idx(x,y,nx)]  = 0.25*( HVt[to_idx(x+1,y,nx)]+HVt[to_idx(x-1,y,nx)]
                                  +HVt[to_idx(x,y+1,nx)]+HVt[to_idx(x,y-1,nx)])
                                  -dt*g*H[to_idx(x,y,nx)]*Zdy[to_idx(x,y,nx)]
       + C*( HUt[to_idx(x,y-1,nx)]*HVt[to_idx(x,y-1,nx)]/Ht[to_idx(x,y-1,nx)]
            -HUt[to_idx(x,y+1,nx)]*HVt[to_idx(x,y+1,nx)]/Ht[to_idx(x,y+1,nx)])
       + C*( pow(HVt[to_idx(x-1,y,nx)],2)/Ht[to_idx(x-1,y,nx)]
            +0.5*g*pow(Ht[to_idx(x-1,y,nx)],2)
            -pow(HVt[to_idx(x+1,y,nx)],2)/Ht[to_idx(x+1,y,nx)]
            -0.5*g*pow(Ht[to_idx(x+1,y,nx)],2));
    }
  }
}

File: compute.m

%% Add data path and start timer
clc; addpath('data/'); tic;

%% Load initial condition from disk

g           = 127267.200000000;	% Gravity, 9.82*(3.6)^2*1000 in [km / hr^2]
Size        = 500;              % Size of map, Size*Size [km]
nx          = 2001;             % Number of cells in each direction on the grid
Tend        = 0.2;             % Simulation time in hours [hr]
dx          = Size/nx;          % Grid spacening

% Set filename
filename = ['Data_nx',num2str(nx),'_',num2str(Size),'km_T',num2str(Tend)]

% Load initial condition from data files
H = fread(fopen([filename,'_h.bin'],'r'),[nx,nx],'double');
HU = fread(fopen([filename,'_hu.bin'],'r'),[nx,nx],'double');
HV = fread(fopen([filename,'_hv.bin'],'r'),[nx,nx],'double');

% Load topography slopes from data files
Zdx = fread(fopen([filename,'_Zdx.bin'],'r'),[nx,nx],'double');
Zdy = fread(fopen([filename,'_Zdy.bin'],'r'),[nx,nx],'double');

% Allocate memory for computing
Ht = zeros(nx,nx);
HUt = zeros(nx,nx);
HVt = zeros(nx,nx);

%save dts
Nmax = 250;
dt_array = zeros(1,Nmax);

%% Compute all time-steps
T = 0; nt = 0;
while nt < Nmax;
    
    %% Compute the time-step length
    
    mu = sqrt( max(abs(HU./H-sqrt(H*g)),abs(HU./H+sqrt(H*g))).^2 + max(abs(HV./H-sqrt(H*g)),abs(HV./H+sqrt(H*g))).^2 );
    dt = dx/(sqrt(2)*max(mu(:)));
    if T+dt > Tend
        dt = Tend-T;
    end
    
    %% Print status
    
    disp(['Computing T: ',num2str(T+dt),'. ',num2str(100*(T+dt)/Tend),'%'])
    
    %% Copy solution to temp storage and enforce boundary condition
    
    Ht = H;
    HUt = HU;
    HVt = HV;
    
    Ht(1,:) = Ht(2,:);
    Ht(end,:) = Ht(end-1,:);
    Ht(:,1) = Ht(:,2);
    Ht(:,end) = Ht(:,end-1);
    
    HUt(1,:) = HUt(2,:);
    HUt(end,:) = HUt(end-1,:);
    HUt(:,1) = HUt(:,2);
    HUt(:,end) = HUt(:,end-1);
    
    HVt(1,:) = HVt(2,:);
    HVt(end,:) = HVt(end-1,:);
    HVt(:,1) = HVt(:,2);
    HVt(:,end) = HVt(:,end-1);
    
    %% Compute a time-step
    
    C = (0.5*dt/dx);
    
    H(2:nx-1,2:nx-1,1) = 0.25*(Ht(2:nx-1,1:nx-2)+Ht(2:nx-1,3:nx)+Ht(1:nx-2,2:nx-1)+Ht(3:nx,2:nx-1)) ...
        + C*( HUt(2:nx-1,1:nx-2) - HUt(2:nx-1,3:nx) + HVt(1:nx-2,2:nx-1) - HVt(3:nx,2:nx-1) );
    
    HU(2:nx-1,2:nx-1) = 0.25*(HUt(2:nx-1,1:nx-2)+HUt(2:nx-1,3:nx)+HUt(1:nx-2,2:nx-1)+HUt(3:nx,2:nx-1)) - dt*g*H(2:nx-1,2:nx-1).*Zdx(2:nx-1,2:nx-1) ...
        + C*( HUt(2:nx-1,1:nx-2).^2./Ht(2:nx-1,1:nx-2) + 0.5*g*Ht(2:nx-1,1:nx-2).^2 - HUt(2:nx-1,3:nx).^2./Ht(2:nx-1,3:nx) - 0.5*g*Ht(2:nx-1,3:nx).^2 ) ...
        + C*( HUt(1:nx-2,2:nx-1).*HVt(1:nx-2,2:nx-1)./Ht(1:nx-2,2:nx-1) - HUt(3:nx,2:nx-1).*HVt(3:nx,2:nx-1)./Ht(3:nx,2:nx-1) );
    
    HV(2:nx-1,2:nx-1) = 0.25*(HVt(2:nx-1,1:nx-2)+HVt(2:nx-1,3:nx)+HVt(1:nx-2,2:nx-1)+HVt(3:nx,2:nx-1)) - dt*g*H(2:nx-1,2:nx-1).*Zdy(2:nx-1,2:nx-1)  ...
        + C*( HUt(2:nx-1,1:nx-2).*HVt(2:nx-1,1:nx-2)./Ht(2:nx-1,1:nx-2) - HUt(2:nx-1,3:nx).*HVt(2:nx-1,3:nx)./Ht(2:nx-1,3:nx) ) ...
        + C*( HVt(1:nx-2,2:nx-1).^2./Ht(1:nx-2,2:nx-1) + 0.5*g*Ht(1:nx-2,2:nx-1).^2 - HVt(3:nx,2:nx-1).^2./Ht(3:nx,2:nx-1) - 0.5*g*Ht(3:nx,2:nx-1).^2  );
    
    %% Impose tolerances
    
    Ht(Ht<0) = 0.00001;
    HUt(Ht<=0.0001) = 0;
    HVt(Ht<=0.0001) = 0;
    
    %% Update time T
    
    T = T + dt;
    nt = nt + 1;
    dt
    dt_array(nt)=Ht(1);
end

%% Save solution to disk

filename = ['output/Matlab_Solution_nx',num2str(nx),'_',num2str(Size),'km','_T',num2str(Tend),'_h.bin'];
fileID = fopen(filename,'w');
fwrite(fileID,Ht,'double');
fclose(fileID);

filename = ['output/Matlab_dt_nx',num2str(nx),'_',num2str(Size),'km','_T',num2str(Tend),'_h.bin'];
fileID = fopen(filename,'w');
fwrite(fileID,dt_array,'double');
fclose(fileID);


% Stop timer
time = toc;

%% Communicate time-to-compute

ops = nt*( 15 + 2 + 11 + 30 + 30 + 1 )*nx^2;
flops = ops/time;
disp(['Time to compute solution : ',num2str(time),' seconds'])
disp(['Average performance      : ',num2str(flops/1.0e9),' gflops'])
